src/ucs/datastruct/mpmc.c

```diff
@@ -59,7 +59,7 @@ ucs_status_t ucs_mpmc_queue_pull(ucs_mpmc_queue_t *mpmc, uint64_t *value_p)
     ucs_status_t status = UCS_ERR_NO_PROGRESS;
     ucs_mpmc_elem_t *elem;
 
-    if (ucs_queue_is_empty(&mpmc->queue)) {
+    if (ucs_queue_is_empty_no_deref(&mpmc->queue)) {
```

Comment:
I doubt this is the best way to fix it.
I see the only usage of `ucs_mpmc_queue_pull` in production code:
```C
async.c
    while (!ucs_mpmc_queue_is_empty(&async->missed)) {
        status = ucs_mpmc_queue_pull(&async->missed, &value);
```
So from there we also check the queue size without taking a mutex, which may end up with the same crash, right?
I think we better change this function:
```C
static inline int ucs_mpmc_queue_is_empty(ucs_mpmc_queue_t *mpmc)
{
    return ucs_queue_is_empty(&mpmc->queue);
}
```
And call it from ucs_mpmc_queue_pull maybe?

Reply:
In addition to that, I'm not sure at all what's the advantage of this micro-optimisation.
It's fragile and error-prone, and also not surer reliable. Accessing shared data without a lock is typically not a good idea. Without acquiring a lock the changes made in one thread might not be visible in another one until some memory fence call occurs. IMO better do it properly with spin lock acquisition, especially if it is not in the hot path

Reply:
please note the intent of that code (meant for busy polling as i understand it) is to check if it _could_ be not empty. doing that, there is _no_ dereference. if it _could_ be not empty, then the code checks again, under the lock this time to confirm. with contention i saw (huge) benefit with this scheme.

also the crash is purely asan checking related, as it implicitly dereferences the `->tail` pointer.

Reply:
Right, I see. Basically what I wanted to say in the last comment is that this check is not reliable, because without acquiring a lock thread might not see the write from another thread, meaning that we may return even if queue is not empty (due to visibility across threads).
That may be totally fine to skip updates, will catch up on them later, if you say performance gain is big

And what do you think about my first comment: since ucs_mpmc_queue_is_empty is actually used to check the queue size, it should be also probably using `no_deref` version

Reply:
ah, you already implemented that)
Maybe just use `ucs_mpmc_queue_is_empty` here?

Reply:
fixed